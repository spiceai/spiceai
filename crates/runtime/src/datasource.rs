use arrow::record_batch::RecordBatch;
use snafu::prelude::*;
use std::future::Future;

#[derive(Debug, Snafu)]
pub enum Error {
    DataSource {},
}

type Result<T, E = Error> = std::result::Result<T, E>;

pub struct DataUpdate {
    pub log_sequence_number: u64,
    pub data: Vec<RecordBatch>,
}

pub trait DataSource {
    fn get_data(&self) -> impl Future<Output = Result<DataUpdate>> + Send;
}
