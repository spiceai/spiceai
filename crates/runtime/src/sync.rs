use snafu::prelude::*;

#[derive(Debug, Snafu)]
#[allow(clippy::enum_variant_names)]
pub enum Error {
    #[snafu(display("Invalid length"))]
    InvalidLength { context: &'static str },

    #[snafu(display("Invalid magic bytes"))]
    InvalidMagicBytes,

    #[snafu(display("Invalid prefix"))]
    InvalidPrefix { source: std::string::FromUtf8Error },

    #[snafu(display("Invalid LSN"))]
    InvalidLogSequenceNumber,
}

type Result<T, E = Error> = std::result::Result<T, E>;

pub mod cursor {
    use super::*;

    const MAGIC_BYTES: [u8; 4] = [83, 80, 83, 67]; // ASCII for "SPSC" (SPice Sync Cursor)

    // Cursor is used to synchronize data between the
    // Spice.ai Runtime and the Spice.ai CLI.
    pub struct Cursor {
        version: u8,
        prefix: String,
        log_sequence_number: u64,
    }

    impl Cursor {
        pub fn parse(cursor: impl AsRef<[u8]>) -> Result<Self> {
            let cursor = cursor.as_ref();
            let mut iter = cursor.iter();

            // Check for magic bytes
            for &magic_byte in &MAGIC_BYTES {
                if iter.next() != Some(&magic_byte) {
                    return InvalidMagicBytesSnafu.fail();
                }
            }

            // Parse version
            let version = *iter.next().ok_or(Error::InvalidLength {
                context: "version: *iter.next().ok_or",
            })?;

            // Parse ASCII string prefix
            let prefix_length = *iter.next().ok_or(Error::InvalidLength {
                context: "prefix_length: *iter.next().ok_or",
            })?;
            let prefix_bytes: Vec<u8> = iter.by_ref().take(prefix_length.into()).copied().collect();
            let prefix = String::from_utf8(prefix_bytes)
                .context(InvalidPrefixSnafu)?
                .trim_matches(char::is_whitespace)
                .to_string();

            // Parse log sequence number (8 bytes, big endian)
            let log_sequence_number_bytes: Vec<u8> = iter.by_ref().take(8).copied().collect();
            if log_sequence_number_bytes.len() != 8 {
                return InvalidLengthSnafu {
                    context: "log_sequence_number_bytes.len() != 8",
                }
                .fail();
            }
            let log_sequence_number_bytes: [u8; 8] = match log_sequence_number_bytes.try_into() {
                Ok(bytes) => bytes,
                Err(_) => return InvalidLogSequenceNumberSnafu.fail(),
            };
            let log_sequence_number = u64::from_be_bytes(log_sequence_number_bytes);

            Ok(Cursor {
                version,
                prefix,
                log_sequence_number,
            })
        }

        pub fn to_bytes(&self) -> Result<Vec<u8>> {
            let mut bytes = Vec::with_capacity(28);

            // Add magic bytes
            bytes.extend_from_slice(&MAGIC_BYTES);

            // Add version byte
            bytes.push(self.version);

            // Add ASCII string prefix
            let prefix_bytes = self.prefix.as_bytes().to_vec();
            let prefix_length = match u8::try_from(prefix_bytes.len()) {
                Ok(length) => length,
                Err(_err) => {
                    return Err(Error::InvalidLength {
                        context: "prefix length > u8::MAX (255)",
                    })
                }
            };
            bytes.push(prefix_length);
            bytes.extend_from_slice(&prefix_bytes);

            // Add log sequence number (8 bytes, big endian)
            let log_sequence_number_bytes = self.log_sequence_number.to_be_bytes();
            bytes.extend_from_slice(&log_sequence_number_bytes);

            Ok(bytes)
        }
    }

    #[cfg(test)]
    #[allow(clippy::expect_used)]
    mod tests {
        use super::*;

        #[test]
        fn test_parse() {
            let cursor_test_bytes: Vec<u8> = vec![
                83, 80, 83, 67, // Magic bytes (SPSC)
                1,  // Version
                12, 115, 112, 105, 99, 101, 46, 97, 105, 46, 99, 108, 105, // Prefix
                255, 255, 255, 255, 255, 255, 0, 0, // Log sequence number
            ];
            let cursor = match Cursor::parse(&cursor_test_bytes) {
                Ok(cursor) => cursor,
                Err(err) => panic!("Unable to parse cursor: {err:?}"),
            };
            assert_eq!(cursor.version, 1);
            assert_eq!(cursor.prefix, "spice.ai.cli");
            assert_eq!(cursor.log_sequence_number, 0xFFFF_FFFF_FFFF_0000_u64);

            let cursor_bytes = cursor
                .to_bytes()
                .expect("Unable to convert cursor to bytes");

            assert_eq!(cursor_bytes, cursor_test_bytes);
        }
    }
}
