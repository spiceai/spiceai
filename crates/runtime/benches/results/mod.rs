use std::sync::Arc;

use arrow::{
    array::{Int32Builder, Int64Builder, RecordBatch, StringBuilder},
    datatypes::{DataType, Field, Schema, SchemaRef},
};
use runtime::dataupdate::{DataUpdate, UpdateType};

#[derive(Copy, Clone)]
pub(crate) enum Status {
    Passed,
}

impl std::fmt::Display for Status {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Status::Passed => write!(f, "passed"),
        }
    }
}

pub(crate) struct BenchmarkResultsBuilder {
    this_run_id: String,
    this_commit_sha: String,
    this_branch_name: String,
    this_iterations: i32,

    run_id: StringBuilder,
    started_at: Int64Builder,
    finished_at: Int64Builder,
    query_name: StringBuilder,
    status: StringBuilder,
    min_duration_ms: Int64Builder,
    max_duration_ms: Int64Builder,
    iterations: Int32Builder,
    commit_sha: StringBuilder,
    branch_name: StringBuilder,
}

impl BenchmarkResultsBuilder {
    pub(crate) fn new(commit_sha: String, branch_name: String, iterations: i32) -> Self {
        Self {
            this_run_id: uuid::Uuid::new_v4().to_string(),
            this_commit_sha: commit_sha,
            this_branch_name: branch_name,
            this_iterations: iterations,
            run_id: StringBuilder::new(),
            started_at: Int64Builder::new(),
            finished_at: Int64Builder::new(),
            query_name: StringBuilder::new(),
            status: StringBuilder::new(),
            min_duration_ms: Int64Builder::new(),
            max_duration_ms: Int64Builder::new(),
            iterations: Int32Builder::new(),
            commit_sha: StringBuilder::new(),
            branch_name: StringBuilder::new(),
        }
    }

    pub(crate) fn record_result(
        &mut self,
        start_time: i64,
        end_time: i64,
        query_name: &str,
        status: Status,
        min_duration_ms: i64,
        max_duration_ms: i64,
    ) {
        self.run_id.append_value(&self.this_run_id);
        self.started_at.append_value(start_time);
        self.finished_at.append_value(end_time);
        self.query_name.append_value(query_name);
        self.status.append_value(status.to_string());
        self.min_duration_ms.append_value(min_duration_ms);
        self.max_duration_ms.append_value(max_duration_ms);
        self.iterations.append_value(self.this_iterations);
        self.commit_sha.append_value(&self.this_commit_sha);
        self.branch_name.append_value(&self.this_branch_name);
    }

    pub(crate) fn iterations(&self) -> i32 {
        self.this_iterations
    }

    pub(crate) fn build(mut self) -> RecordBatch {
        let schema = results_schema();
        let batch = RecordBatch::try_new(
            schema,
            vec![
                Arc::new(self.run_id.finish()),
                Arc::new(self.started_at.finish()),
                Arc::new(self.finished_at.finish()),
                Arc::new(self.query_name.finish()),
                Arc::new(self.status.finish()),
                Arc::new(self.min_duration_ms.finish()),
                Arc::new(self.max_duration_ms.finish()),
                Arc::new(self.iterations.finish()),
                Arc::new(self.commit_sha.finish()),
                Arc::new(self.branch_name.finish()),
            ],
        );
        match batch {
            Ok(batch) => batch,
            Err(e) => panic!("Error building record batch: {e}"),
        }
    }
}

impl From<BenchmarkResultsBuilder> for DataUpdate {
    fn from(builder: BenchmarkResultsBuilder) -> Self {
        let batch = builder.build();
        DataUpdate {
            schema: batch.schema(),
            data: vec![batch],
            update_type: UpdateType::Append,
        }
    }
}

fn results_schema() -> SchemaRef {
    let fields = vec![
        Field::new("run_id", DataType::Utf8, false),
        Field::new("started_at", DataType::Int64, false),
        Field::new("finished_at", DataType::Int64, false),
        Field::new("query_name", DataType::Utf8, false),
        Field::new("status", DataType::Utf8, false),
        Field::new("min_duration_ms", DataType::Int64, false),
        Field::new("max_duration_ms", DataType::Int64, false),
        Field::new("iterations", DataType::Int32, false),
        Field::new("commit_sha", DataType::Utf8, false),
        Field::new("branch_name", DataType::Utf8, false),
    ];
    Arc::new(Schema::new(fields))
}
