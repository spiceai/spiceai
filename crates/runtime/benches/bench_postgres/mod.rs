/*
Copyright 2024 The Spice.ai OSS Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
     https://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use app::AppBuilder;
use runtime::Runtime;

use crate::results::BenchmarkResultsBuilder;
use spicepod::component::{dataset::Dataset, params::Params};

pub(crate) async fn run(
    rt: &mut Runtime,
    benchmark_results: &mut BenchmarkResultsBuilder,
) -> Result<(), String> {
    let test_queries = get_test_queries();

    for (query_name, query) in test_queries {
        super::run_query_and_record_result(rt, benchmark_results, "postgres", query_name, query)
            .await?;
    }

    Ok(())
}

pub fn build_app(app_builder: AppBuilder) -> AppBuilder {
    app_builder
        .with_dataset(make_dataset("customer", "customer"))
        .with_dataset(make_dataset("lineitem", "lineitem"))
        .with_dataset(make_dataset("part", "part"))
        .with_dataset(make_dataset("partsupp", "partsupp"))
        .with_dataset(make_dataset("orders", "orders"))
        .with_dataset(make_dataset("nation", "nation"))
        .with_dataset(make_dataset("region", "region"))
        .with_dataset(make_dataset("supplier", "supplier"))
}

fn get_postgres_params() -> Params {
    let pg_host = std::env::var("PG_BENCHMARK_PG_HOST").unwrap_or_default();
    let pg_user = std::env::var("PG_BENCHMARK_PG_USER").unwrap_or_default();
    let pg_pass = std::env::var("PG_BENCHMARK_PG_PASS").unwrap_or_default();
    let pg_db = std::env::var("PG_BENCHMARK_PG_DB").unwrap_or_default();
    let pg_sslmode = std::env::var("PG_BENCHMARK_PG_SSLMODE").unwrap_or_default();
    Params::from_string_map(
        vec![
            ("pg_host".to_string(), pg_host),
            ("pg_user".to_string(), pg_user),
            ("pg_db".to_string(), pg_db),
            ("pg_pass".to_string(), pg_pass),
            ("pg_sslmode".to_string(), pg_sslmode),
        ]
        .into_iter()
        .collect(),
    )
}

fn make_dataset(path: &str, name: &str) -> Dataset {
    let mut dataset = Dataset::new(format!("postgres:{path}"), name.to_string());
    dataset.params = Some(get_postgres_params());
    dataset
}

fn get_test_queries() -> Vec<(&'static str, &'static str)> {
    vec![
        ("tpch_q1", include_str!("../queries/tpch_q1.sql")),
        ("tpch_q2", include_str!("../queries/tpch_q2.sql")),
        ("tpch_q3", include_str!("../queries/tpch_q3.sql")),
        ("tpch_q4", include_str!("../queries/tpch_q4.sql")),
        ("tpch_q5", include_str!("../queries/tpch_q5.sql")),
        ("tpch_q6", include_str!("../queries/tpch_q6.sql")),
        ("tpch_q7", include_str!("../queries/tpch_q7.sql")),
        ("tpch_q8", include_str!("../queries/tpch_q8.sql")),
        ("tpch_q9", include_str!("../queries/tpch_q9.sql")),
        ("tpch_q10", include_str!("../queries/tpch_q10.sql")),
        ("tpch_q11", include_str!("../queries/tpch_q11.sql")),
        ("tpch_q12", include_str!("../queries/tpch_q12.sql")),
        ("tpch_q13", include_str!("../queries/tpch_q13.sql")),
        ("tpch_q14", include_str!("../queries/tpch_q14.sql")),
        // tpch_q15 has a view creation which we don't support by design
        ("tpch_q16", include_str!("../queries/tpch_q16.sql")),
        ("tpch_q17", include_str!("../queries/tpch_q17.sql")),
        ("tpch_q18", include_str!("../queries/tpch_q18.sql")),
        ("tpch_q19", include_str!("../queries/tpch_q19.sql")),
        ("tpch_q20", include_str!("../queries/tpch_q20.sql")),
        ("tpch_q21", include_str!("../queries/tpch_q21.sql")),
        ("tpch_q22", include_str!("../queries/tpch_q22.sql")),
        (
            "tpch_simple_q1",
            include_str!("../queries/tpch_simple_q1.sql"),
        ),
        (
            "tpch_simple_q2",
            include_str!("../queries/tpch_simple_q2.sql"),
        ),
        (
            "tpch_simple_q3",
            include_str!("../queries/tpch_simple_q3.sql"),
        ),
        (
            "tpch_simple_q4",
            include_str!("../queries/tpch_simple_q4.sql"),
        ),
        (
            "tpch_simple_q5",
            include_str!("../queries/tpch_simple_q5.sql"),
        ),
    ]
}
