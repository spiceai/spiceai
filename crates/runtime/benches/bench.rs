/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

//! This is a benchmark test suite for the Spice runtime.
//!
//! It performs the following actions:
//! 1. Starts the runtime with all of the datasets to test loaded.
//! 2. Runs a series of queries against the runtime.
//! 3. Reports the results to the spice.ai dataset <https://spice.ai/spicehq/spice-tests/datasets/oss_benchmarks>

// spice.ai/spicehq/spice-tests/datasets/spicehq."spice-tests".oss_benchmarks
// schema
// run_id, started_at, finished_at, connector_name, query_name, status, min_duration, max_duration, iterations, commit_sha

use std::panic;
use std::sync::Arc;

#[cfg(feature = "postgres")]
use crate::bench_postgres::get_postgres_params;
use crate::results::Status;
use arrow::array::RecordBatch;
use clap::Parser;
use datafusion::datasource::provider_as_source;
use datafusion::logical_expr::{LogicalPlanBuilder, UNNAMED_TABLE};
use datafusion::{dataframe::DataFrame, datasource::MemTable, execution::context::SessionContext};
use results::BenchmarkResultsBuilder;
use runtime::{dataupdate::DataUpdate, Runtime};
use spicepod::component::dataset::acceleration::{self, Acceleration, Mode};
use spicepod::component::params::Params;

mod results;
mod setup;

mod bench_s3;
mod bench_spicecloud;

#[cfg(feature = "delta_lake")]
mod bench_delta;
#[cfg(feature = "mysql")]
mod bench_mysql;
#[cfg(feature = "odbc")]
mod bench_odbc_athena;
#[cfg(feature = "odbc")]
mod bench_odbc_databricks;
#[cfg(feature = "postgres")]
mod bench_postgres;
#[cfg(feature = "spark")]
mod bench_spark;

// Define command line arguments for running benchmark test
#[derive(Parser, Debug)]
#[command(version, about, long_about = None)]
struct BenchArgs {
    /// Run the benchmark
    #[arg(long)]
    bench: bool,

    /// Set the connector to run benchmark test on
    #[arg(short, long)]
    connector: Option<String>,

    /// Set the accelerator to run benchmark test on
    #[arg(short, long)]
    accelerator: Option<String>,

    /// Set the acceleration mode for accelerator
    #[arg(short, long)]
    mode: Option<String>,

    /// Set the benchmark to run: TPCH / TPCDS
    #[arg(short, long, default_value = "tpch")]
    bench_name: String,
}

#[tokio::main]
async fn main() -> Result<(), String> {
    let _ = rustls::crypto::CryptoProvider::install_default(
        rustls::crypto::aws_lc_rs::default_provider(),
    );

    let mut upload_results_dataset: Option<String> = None;
    if let Ok(env_var) = std::env::var("UPLOAD_RESULTS_DATASET") {
        println!("UPLOAD_RESULTS_DATASET: {env_var}");
        upload_results_dataset = Some(env_var);
    }

    let args = BenchArgs::parse();

    match (args.connector.as_deref(), args.accelerator.as_deref(), args.mode.as_deref()) {
        (None, None, None) => {
            // Run all connector / accelerator benchmark tests
            let connectors = vec![
                "spice.ai",
                "s3",
                #[cfg(feature = "spark")]
                "spark",
                #[cfg(feature = "postgres")]
                "postgres",
                #[cfg(feature = "mysql")]
                "mysql",
                #[cfg(feature = "odbc")]
                "odbc-databricks",
                #[cfg(feature = "odbc")]
                "odbc-athena",
                #[cfg(feature = "delta_lake")]
                "delta_lake",
            ];
            for connector in connectors {
                run_connector_bench(connector, &upload_results_dataset, args.bench_name.as_ref()).await?;
            }
            let accelerators: Vec<Acceleration> = vec![
                create_acceleration("arrow", acceleration::Mode::Memory, None),
                #[cfg(feature = "duckdb")]
                create_acceleration("duckdb", acceleration::Mode::Memory, None),
                #[cfg(feature = "duckdb")]
                create_acceleration("duckdb", acceleration::Mode::File, None),
                #[cfg(feature = "sqlite")]
                create_acceleration("sqlite", acceleration::Mode::Memory, None),
                #[cfg(feature = "sqlite")]
                create_acceleration("sqlite", acceleration::Mode::File, None),
                #[cfg(feature = "postgres")]
                create_acceleration("postgres", acceleration::Mode::Memory, Some(get_postgres_params(true))),
            ];
            for accelerator in accelerators {
                run_accelerator_bench(accelerator.clone(), &upload_results_dataset, "tpch").await?;
                run_accelerator_bench(accelerator, &upload_results_dataset, "tpcds").await?;
            }
        },
        (Some(connector), None, None) => {
            // Run connector benchmark test
            run_connector_bench(connector, &upload_results_dataset, args.bench_name.as_ref()).await?;
        },
        (None, Some(accelerator), mode) => {
            // Run accelerator benchmark test
            let mode = match mode {
                Some("file") => Mode::File,
                Some("memory") | None => Mode::Memory,
                _ => return Err(format!("Invalid mode parameter for {accelerator} accelerator")),
            };

            let params: Option<Params> = {
                #[cfg(feature = "postgres")]
                {
                    if accelerator == "postgres" {
                        Some(get_postgres_params(true))
                    } else {
                        None
                    }
                }
                #[cfg(not(feature = "postgres"))]
                {
                    None
                }
            };

            let acceleration = create_acceleration(accelerator, mode, params);

            match args.bench_name.as_ref() {
                "tpch" => {
                    run_accelerator_bench(acceleration, &upload_results_dataset, "tpch").await?;
                }
                "tpcds" => {
                    run_accelerator_bench(acceleration, &upload_results_dataset, "tpcds").await?;
                }
                _ => return Err(format!("Invalid mode bench_name parameter {}", args.bench_name)),
            }
        },
        _ => return Err("Invalid command line input: accelerator or mode parameter supplied for connector benchmark".to_string()),
    }

    Ok(())
}

async fn run_connector_bench(
    connector: &str,
    upload_results_dataset: &Option<String>,
    bench_name: &str,
) -> Result<(), String> {
    // TODO: Implement and enable connector TPCDS bench if it's required
    if bench_name == "tpcds" {
        return Err("TPCDS Benchmark not implemented for data connectors".to_string());
    }

    let mut display_records = vec![];

    let (mut benchmark_results, mut rt) =
        setup::setup_benchmark(upload_results_dataset, connector, None, bench_name).await;

    match connector {
        "spice.ai" => {
            bench_spicecloud::run(&mut rt, &mut benchmark_results).await?;
        }
        "s3" => {
            bench_s3::run(&mut rt, &mut benchmark_results, None, None, "tpch").await?;
        }
        #[cfg(feature = "spark")]
        "spark" => {
            bench_spark::run(&mut rt, &mut benchmark_results).await?;
        }
        #[cfg(feature = "postgres")]
        "postgres" => {
            bench_postgres::run(&mut rt, &mut benchmark_results).await?;
        }
        #[cfg(feature = "mysql")]
        "mysql" => {
            bench_mysql::run(&mut rt, &mut benchmark_results).await?;
        }
        #[cfg(feature = "odbc")]
        "odbc-databricks" => {
            bench_odbc_databricks::run(&mut rt, &mut benchmark_results).await?;
        }
        #[cfg(feature = "odbc")]
        "odbc-athena" => {
            bench_odbc_athena::run(&mut rt, &mut benchmark_results).await?;
        }
        #[cfg(feature = "delta_lake")]
        "delta_lake" => {
            bench_delta::run(&mut rt, &mut benchmark_results).await?;
        }
        _ => {}
    }
    let data_update: DataUpdate = benchmark_results.into();

    let mut records = data_update.data.clone();
    display_records.append(&mut records);

    if let Some(upload_results_dataset) = upload_results_dataset.clone() {
        tracing::info!("Writing benchmark results to dataset {upload_results_dataset}...");
        setup::write_benchmark_results(data_update, &rt).await?;
    }

    display_benchmark_records(display_records).await?;
    Ok(())
}

async fn run_accelerator_bench(
    accelerator: Acceleration,
    upload_results_dataset: &Option<String>,
    bench_name: &str,
) -> Result<(), String> {
    let mut display_records = vec![];

    let engine = accelerator.engine.clone();
    let mode = accelerator.mode.clone();

    let (mut benchmark_results, mut rt) =
        setup::setup_benchmark(upload_results_dataset, "s3", Some(accelerator), bench_name).await;

    bench_s3::run(
        &mut rt,
        &mut benchmark_results,
        engine,
        Some(mode),
        bench_name,
    )
    .await?;

    let data_update: DataUpdate = benchmark_results.into();

    let mut records = data_update.data.clone();
    display_records.append(&mut records);

    if let Some(upload_results_dataset) = upload_results_dataset.clone() {
        tracing::info!("Writing benchmark results to dataset {upload_results_dataset}...");
        setup::write_benchmark_results(data_update, &rt).await?;
    }

    display_benchmark_records(display_records).await?;
    Ok(())
}

fn create_acceleration(
    engine: &str,
    mode: acceleration::Mode,
    params: Option<Params>,
) -> Acceleration {
    Acceleration {
        engine: Some(engine.to_string()),
        mode,
        params,
        ..Default::default()
    }
}

fn get_current_unix_ms() -> i64 {
    let now = std::time::SystemTime::now();
    now.duration_since(std::time::UNIX_EPOCH)
        .map(|d| i64::try_from(d.as_millis()).unwrap_or(0))
        .unwrap_or(0)
}

async fn run_query_and_record_result(
    rt: &mut Runtime,
    benchmark_results: &mut BenchmarkResultsBuilder,
    connector: &str,
    query_name: &str,
    query: &str,
    verify_query_result: bool,
) -> Result<(), String> {
    // Additional round of query run before recording results.
    // To discard the abnormal results caused by: establishing initial connection / spark cluster startup time
    let _ = run_query(rt, connector, query_name, query).await;
    let snapshot_err = record_explain_plan(rt, connector, query_name, query)
        .await
        .err();

    tracing::info!("Running query `{connector}` `{query_name}`...");
    let start_time = get_current_unix_ms();

    let mut min_iter_duration_ms = i64::MAX;
    let mut max_iter_duration_ms = i64::MIN;

    let mut query_err: Option<String> = None;

    let mut completed_iterations = 0;

    for idx in 0..benchmark_results.iterations() {
        completed_iterations += 1;

        let start_iter_time = get_current_unix_ms();

        let res = run_query(rt, connector, query_name, query).await;
        let end_iter_time = get_current_unix_ms();

        let iter_duration_ms = end_iter_time - start_iter_time;
        if iter_duration_ms < min_iter_duration_ms {
            min_iter_duration_ms = iter_duration_ms;
        }
        if iter_duration_ms > max_iter_duration_ms {
            max_iter_duration_ms = iter_duration_ms;
        }

        match res {
            Ok(records) => {
                // Show the first 10 records of the result from the first iteration
                if idx == 0 {
                    let num_rows = records
                        .iter()
                        .map(arrow::array::RecordBatch::num_rows)
                        .sum::<usize>();

                    let limited_records: Vec<_> = records
                        .iter()
                        .flat_map(|batch: &RecordBatch| {
                            (0..batch.num_rows()).map(move |i| batch.slice(i, 1))
                        })
                        .take(10)
                        .collect();

                    let records_pretty =
                        arrow::util::pretty::pretty_format_batches(&limited_records)
                            .map_err(|e| e.to_string())?;

                    tracing::info!(
                    "Query `{connector}` `{query_name}` returned {num_rows} rows:\n{records_pretty}",
                );
                    if verify_query_result {
                        let result = panic::catch_unwind(|| {
                            insta::assert_snapshot!(
                                format!("{connector}_{query_name}"),
                                records_pretty
                            );
                        });
                        if result.is_err() {
                            let error_str = format!(
                                "Query `{connector}` `{query_name}` snapshot assertion failed",
                            );
                            tracing::error!(error_str);
                            query_err = Some(error_str);
                            break;
                        }
                    }
                }
            }
            Err(e) => {
                tracing::error!(
                    "Query `{connector}` `{query_name}` iteration {idx} failed with error: \n{e}",
                );
                query_err = Some(e);
                break;
            }
        }
    }

    let end_time = get_current_unix_ms();

    // Both query failure and snapshot test failure will cause the result to be written as Status::Failed
    benchmark_results.record_result(
        start_time,
        end_time,
        connector,
        query_name,
        if query_err.is_some() || snapshot_err.is_some() {
            Status::Failed
        } else {
            Status::Passed
        },
        min_iter_duration_ms,
        max_iter_duration_ms,
        completed_iterations,
    );

    match (query_err, snapshot_err) {
        (Some(query), Some(snapshot)) => {
            return Err(format!(
                "Query Error: {query}; Snapshot Test Error: {snapshot}",
            ));
        }
        (Some(query), None) => {
            return Err(format!("Query Error: {query}"));
        }
        (None, Some(snapshot)) => {
            return Err(format!("Snapshot Test Error: {snapshot}"));
        }
        (None, None) => {}
    }

    Ok(())
}

async fn run_query(
    rt: &mut Runtime,
    connector: &str,
    query_name: &str,
    query: &str,
) -> Result<Vec<RecordBatch>, String> {
    let res = rt
        .datafusion()
        .ctx
        .sql(query)
        .await
        .map_err(|e| format!("query `{connector}` `{query_name}` to plan: {e}"))?
        .collect()
        .await
        .map_err(|e| format!("query `{connector}` `{query_name}` to results: {e}"))?;

    Ok(res)
}

const ENABLED_SNAPSHOT_CONNECTORS: &[&str] = &["spice.ai", "s3", "s3_arrow_memory"];

async fn record_explain_plan(
    rt: &mut Runtime,
    connector: &str,
    query_name: &str,
    query: &str,
) -> Result<(), String> {
    // TODO: Turn on snapshot for tpcds queries after tpcds hardening
    if query_name.starts_with("tpcds") {
        return Ok(());
    }

    if !ENABLED_SNAPSHOT_CONNECTORS.contains(&connector) {
        return Ok(());
    }

    // Check the plan
    let plan_results = rt
        .datafusion()
        .ctx
        .sql(&format!("EXPLAIN {query}"))
        .await
        .map_err(|e| format!("query `{query}` to plan: {e}"))?
        .collect()
        .await
        .map_err(|e| format!("query `{query}` to results: {e}"))?;

    let Ok(explain_plan) = arrow::util::pretty::pretty_format_batches(&plan_results) else {
        return Err("Failed to format plan".to_string());
    };

    let mut assertion_err: Option<String> = None;

    insta::with_settings!({
        description => format!("Query: {query}"),
        omit_expression => true,
        filters => vec![
            (r"required_guarantees=\[[^\]]*\]", "required_guarantees=[N]"),
        ],
    }, {
        let result = panic::catch_unwind(|| {
            insta::assert_snapshot!(format!("{connector}_{query_name}_explain"), explain_plan);
        });
        if result.is_err() {
            assertion_err = Some(format!("Snapshort assertion failed for {connector}, {query_name}"));
        }
    });

    if let Some(assertion_err) = assertion_err {
        return Err(assertion_err);
    }

    Ok(())
}

/// Display the benchmark results record batches to the console.
async fn display_benchmark_records(records: Vec<RecordBatch>) -> Result<(), String> {
    if records.is_empty() {
        return Ok(());
    }

    let schema = records[0].schema();

    let ctx = SessionContext::new();
    let provider = MemTable::try_new(schema, vec![records]).map_err(|e| e.to_string())?;
    let df = DataFrame::new(
        ctx.state(),
        LogicalPlanBuilder::scan(UNNAMED_TABLE, provider_as_source(Arc::new(provider)), None)
            .map_err(|e| e.to_string())?
            .build()
            .map_err(|e| e.to_string())?,
    );

    if let Err(e) = df.show().await {
        println!("Error displaying results: {e}");
    };
    Ok(())
}
