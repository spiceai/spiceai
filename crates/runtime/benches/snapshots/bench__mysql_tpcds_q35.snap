---
source: crates/runtime/benches/bench.rs
assertion_line: 386
expression: records_pretty
---
+----------+-----------+-------------------+--------------+------+-----------------------------------------+--------------------+--------------------+-----------------------+------+--------------------------------------------------+-----------------------------+-----------------------------+----------------------+------+-------------------------------------------------+----------------------------+----------------------------+
| ca_state | cd_gender | cd_marital_status | cd_dep_count | cnt1 | max(customer_demographics.cd_dep_count) | stddev_dep_count_1 | stddev_dep_count_2 | cd_dep_employed_count | cnt2 | max(customer_demographics.cd_dep_employed_count) | stddev_dep_employed_count_1 | stddev_dep_employed_count_2 | cd_dep_college_count | cnt3 | max(customer_demographics.cd_dep_college_count) | stddev_dep_college_count_1 | stddev_dep_college_count_2 |
+----------+-----------+-------------------+--------------+------+-----------------------------------------+--------------------+--------------------+-----------------------+------+--------------------------------------------------+-----------------------------+-----------------------------+----------------------+------+-------------------------------------------------+----------------------------+----------------------------+
|          | F         | D                 | 0            | 1    | 0                                       | 0.0                | 0.0                | 1                     | 1    | 1                                                | 0.0                         | 0.0                         | 5                    | 1    | 5                                               | 0.0                        | 0.0                        |
|          | F         | D                 | 0            | 1    | 0                                       | 0.0                | 0.0                | 4                     | 1    | 4                                                | 0.0                         | 0.0                         | 2                    | 1    | 2                                               | 0.0                        | 0.0                        |
|          | F         | D                 | 0            | 1    | 0                                       | 0.0                | 0.0                | 5                     | 1    | 5                                                | 0.0                         | 0.0                         | 2                    | 1    | 2                                               | 0.0                        | 0.0                        |
|          | F         | D                 | 2            | 1    | 2                                       | 0.0                | 0.0                | 1                     | 1    | 1                                                | 0.0                         | 0.0                         | 0                    | 1    | 0                                               | 0.0                        | 0.0                        |
|          | F         | D                 | 2            | 1    | 2                                       | 0.0                | 0.0                | 5                     | 1    | 5                                                | 0.0                         | 0.0                         | 4                    | 1    | 4                                               | 0.0                        | 0.0                        |
|          | F         | D                 | 3            | 1    | 3                                       | 0.0                | 0.0                | 2                     | 1    | 2                                                | 0.0                         | 0.0                         | 0                    | 1    | 0                                               | 0.0                        | 0.0                        |
|          | F         | D                 | 3            | 1    | 3                                       | 0.0                | 0.0                | 4                     | 1    | 4                                                | 0.0                         | 0.0                         | 4                    | 1    | 4                                               | 0.0                        | 0.0                        |
|          | F         | D                 | 3            | 1    | 3                                       | 0.0                | 0.0                | 5                     | 1    | 5                                                | 0.0                         | 0.0                         | 6                    | 1    | 6                                               | 0.0                        | 0.0                        |
|          | F         | D                 | 4            | 1    | 4                                       | 0.0                | 0.0                | 0                     | 1    | 0                                                | 0.0                         | 0.0                         | 5                    | 1    | 5                                               | 0.0                        | 0.0                        |
|          | F         | D                 | 4            | 1    | 4                                       | 0.0                | 0.0                | 1                     | 1    | 1                                                | 0.0                         | 0.0                         | 3                    | 1    | 3                                               | 0.0                        | 0.0                        |
+----------+-----------+-------------------+--------------+------+-----------------------------------------+--------------------+--------------------+-----------------------+------+--------------------------------------------------+-----------------------------+-----------------------------+----------------------+------+-------------------------------------------------+----------------------------+----------------------------+
